sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala

```diff
@@ -133,8 +133,19 @@ object TextInputCSVDataSource extends CSVDataSource {
       sparkSession: SparkSession,
       inputPaths: Seq[FileStatus],
       parsedOptions: CSVOptions): Option[StructType] = {
-    val csv: Dataset[String] = createBaseDataset(sparkSession, inputPaths, parsedOptions)
-    val firstLine: String = CSVUtils.filterCommentAndEmpty(csv, parsedOptions).first()
+    val csv = createBaseDataset(sparkSession, inputPaths, parsedOptions)
+    CSVUtils.filterCommentAndEmpty(csv, parsedOptions)
+      .take(1)
+      .headOption
+      .map(firstLine => infer(sparkSession, parsedOptions, csv, firstLine))
+      .orElse(Some(StructType(Seq())))
```

Comment:
I would suggest that we use pattern matching in order to make it more verbose and avoid code like this:

        if (maybeFirstRow.isDefined) {
            val firstRow = maybeFirstRow.get

I also touched `WholeFileCSVDataSource` to unify both implementations. What's your opinion?
Regarding code de-duplication, I fully agree, that it should be done in separate PR.